### PR TITLE
Xcode 14.3 beta2 build fixes

### PR DESCRIPTION
**Pull request by pepone**
With Xcode 14.3 beta2 the C++17 deprecated features produce warnings even when building with `-std=c++11` mode, this PR updates macOS builds to use `-std=c++17`. There is also a few fixes for building in C++17 mode, and for new warnings that are enabled by default.

### CODE REVIEW

File: cpp/config/Make.rules [hunk 1]
```diff
@@ -99,8 +99,14 @@ endif
 #
 cpp11_cppflags          = -DICE_CPP11_MAPPING
 ifneq ($(shell $(CC) -E config/cplusplus_check.cpp 1> /dev/null 2>&1; echo $$?),0)
+ifeq ($(os),Darwin)
+# With Xcode 14.3 C++17 deprecations apply when building with -std=c++11, we pass -std=c++17
```
Comment (bernardnormier) [inline cpp/config/Make.rules:103]: What is the default -std with the clang versions we support on macos? 
If it is c++17 or greater, I would simply remove the -std flag.

The main reason we have this `=std=c++11` is old compiler's default std was < c++11. Must be rare now.

File: cpp/src/Ice/ConnectionI.cpp [hunk 2]
```diff
@@ -1083,7 +1083,7 @@ Ice::ConnectionI::setCloseCallback(ICE_IN(ICE_DELEGATE(CloseCallback)) callback)
                 CallbackWorkItem(const ConnectionIPtr& connection, ICE_IN(ICE_DELEGATE(CloseCallback)) callback) :
                     _connection(connection),
 #ifdef ICE_CPP11_MAPPING
-                    _callback(move(callback))
+                    _callback(std::move(callback))
```
Comment (bernardnormier) [inline cpp/src/Ice/ConnectionI.cpp:1086]: That's fine but why do we need this `std::`? There is ambiguity with another move function?

File: cpp/src/IceUtil/StringConverter.cpp [hunk 3]
```diff
@@ -2,17 +2,19 @@
...
-
 #include <IceUtil/StringConverter.h>
 #include <IceUtil/MutexPtrLock.h>
 #include <IceUtil/Mutex.h>
 #include <IceUtil/StringUtil.h>
 
+// TODO codecvt was deprecated in C++17 and cause build failures in C++17 mode
+#if (ICE_CPLUSPLUS >= 201703L)
+#   include <IceUtil/DisableWarnings.h>
+#elif defined(_MSC_VER) && (_MSVC_LANG >= 201703L)
```
Comment (bernardnormier) [inline cpp/src/IceUtil/StringConverter.cpp:13]: Is this now dead code?

I assume we set ICE_CPLUSPLUS when defined(_MSC_VER) && (_MSVC_LANG >= 201703L)

File: cpp/config/Make.rules (hunk 1 above)
Comment (pepone) [reply cpp/config/Make.rules:103]: The `__cplusplus` macro defaults to `199711L`
```
echo | clang++ -dM -E -x c++ - | grep cplus
#define __cplusplus 199711L
```

File: cpp/src/IceUtil/StringConverter.cpp (hunk 3 above)
Comment (pepone) [reply cpp/src/IceUtil/StringConverter.cpp:13]: right we can switch the condition the other way around

Comment (pepone) [reply cpp/src/IceUtil/StringConverter.cpp]: I reworked the check

File: cpp/src/Ice/ConnectionI.cpp (hunk 2 above)
Comment (bernardnormier) [reply cpp/src/Ice/ConnectionI.cpp:1086]: I would keep it, I find the `std::move` more readable.

File: cpp/config/Make.rules (hunk 1 above)
Comment (bernardnormier) [reply cpp/config/Make.rules:103]: Ok.

**Review by bernardnormier — APPROVED** [review]
Looks good.

**Review by externl — APPROVED** [review]
Looks good!

**Pull request merged**